examples/sst2/train.py

```diff
@@ -0,0 +1,272 @@
+
+    metrics, model_state = eval_step(model, params, model_state, batch, rngs)
+    batch_metrics.append(metrics)
+
+  batch_metrics = jax.device_get(batch_metrics)
+  metrics = normalize_batch_metrics(batch_metrics)
+  logging.info('eval  epoch %03d loss %.4f accuracy %.2f', epoch,
+               metrics['loss'], metrics['accuracy'] * 100)
+  return metrics, model_state
+
+
+def train_epoch(model, optimizer, model_state, train_batches, epoch, rngs):
+  """Train for a single epoch."""
+  batch_metrics = []
+  for batch in train_batches:
+    # pylint: disable=protected-access
+    batch = jax.tree_map(lambda x: x._numpy(), batch)
+    optimizer, metrics, model_state = train_step(
+        model, optimizer, model_state, batch, rngs)
+    batch_metrics.append(metrics)
+
+  # Compute the metrics for this epoch.
+  batch_metrics = jax.device_get(batch_metrics)
+  metrics = normalize_batch_metrics(batch_metrics)
+
+  logging.info('train epoch %03d loss %.4f accuracy %.2f', epoch,
+               metrics['loss'], metrics['accuracy'] * 100)
+
+  return optimizer, metrics, model_state
+
+
+def model_from_config(config: ml_collections.ConfigDict):
+  """Builds a text classification model from a config."""
+  model = models.TextClassifier(
+      embedding_size=config.embedding_size,
+      hidden_size=config.hidden_size,
+      vocab_size=config.vocab_size,
+      output_size=config.output_size,
+      dropout_rate=config.dropout_rate,
+      word_dropout_rate=config.word_dropout_rate,
+      unk_idx=config.unk_idx,
+      deterministic=config.deterministic)
+  return model
+
+
+def train_and_evaluate(config: ml_collections.ConfigDict, workdir: str):
+  """Execute model training and evaluation loop.
+  Args:
+    config: Hyperparameter configuration for training and evaluation.
+    workdir: Directory where the tensorboard summaries are written to.
+  Returns:
+    The trained optimizer.
+  """
+  # Prepare datasets.
+  train_dataset = input_pipeline.TextDataset(
+      tfds_name='glue/sst2', split='train')
+  eval_dataset = input_pipeline.TextDataset(
+      tfds_name='glue/sst2', split='validation')
+  train_batches = train_dataset.get_bucketed_batches(
+      config.batch_size,
+      config.bucket_size,
+      max_input_length=config.max_input_length,
+      drop_remainder=True,
+      shuffle=True,
+      shuffle_seed=config.seed)
+  eval_batches = eval_dataset.get_batches(batch_size=config.batch_size)
+
+  # Prepare configs.
+  config.vocab_size = len(train_dataset.vocab)
+  eval_config = copy.deepcopy(config)
+  eval_config.deterministic = True
+
+  # Create models.
+  train_model = model_from_config(config)
+  eval_model = model_from_config(eval_config)
+
+  # Initialize parameters.
+  rng = jax.random.PRNGKey(config.seed)
+  params, model_state = get_initial_params_and_state(rng, eval_model)
+  model_state, _ = model_state.pop('intermediates')  # Remove for training.
```

Comment:
Why do we remove these for training? I guess you want to do the analysis at eval so we can remove it for efficiency?

Reply:
It's really important to remove them. During training, the intermediates would be concatenated with the previous ones at each step, triggering a new trace for each step, making things veeeery slow :D

Reply:
Oh yes I actually heard of that as well. That is quite an important detail of the intermediates! Do you want to add a comment here explaining that? It seems very useful for others if they want to use a similar pattern.

Reply:
I added a comment explainig what happens if we don't pop out intermedaites.